.evergreen/scripts/find-cmake-version.sh

```diff
@@ -0,0 +1,240 @@
+#!/usr/bin/env bash
+
+# Used to workaround curl certificate verification failures on certain distros.
+: "${distro_id:?"missing required Evergreen expansion"}"
+
+# Create a temporary directory in the existing directory $1.
+make_tmpdir_in() {
+  declare -r target_dir="${1:?"missing path to target directory"}"
+
+  if [[ ! -d "${target_dir}" ]]; then
+    echo "make_tmpdir_in: target directory ${target_dir} does not exist!" 1>&2
+    return 1
+  fi
+
+  declare tmpdir
+
+  {
+    if [[ "${OSTYPE:?}" == darwin* ]]; then
+      # mktemp on MacOS does not support -p.
+      tmpdir="$(mktemp -d)" || return
+      declare tmpbase
+      tmpbase="$(basename "${tmpdir}")" || return
+      mv -f "${tmpdir}" "${target_dir}" || return
+      tmpdir="${target_dir}/${tmpbase}" || return
+    else
+      tmpdir="$(mktemp -d -p "${target_dir}")" || return
+    fi
+  } 1>&2
+
+  printf "%s" "${tmpdir}"
+}
+
+# Ensure "good enough" atomicity when two or more tasks running on the same host
+# attempt to install a version of CMake.
+cmake_replace_latest() {
+  declare -r cache="${1:?"missing path to cache directory"}"
+  declare -r root="${2:?"missing path to latest CMake root directory"}"
+  declare -r version="${3:?"missing latest CMake version"}"
+  declare -r cmake_cache="${cache}/cmake-${version}"
+
+  # Doesn't matter who creates the cache directory.
+  mkdir -p "${cache}" || return
+
+  # Should not happen, but handle it anyways.
+  [[ -L "${cmake_cache}" ]] || rm -rf "${cmake_cache}" || return
+
+  # Temporary directory to create new symlink.
+  declare tmpdir
+  tmpdir="$(make_tmpdir_in "${cache}")" || return
+
+  # Symlink to the new CMake root directory.
+  ln -s "${root}" "${tmpdir}/cmake-${version}" || return
+
+  # Atomic substitution of symlink from old to new CMake root directory.
+  mv -f "${tmpdir}/cmake-${version}" "${cache}" || return
+
+  # Be nice and clean up the now-empty temporary directory.
+  rmdir "${tmpdir}" || return
+}
+
+# find_cmake_version
+#
+# Usage:
+#   find_cmake_version 3 1 0
+#   cmake_binary="$(find_cmake_version 3 1 0)"
+#   cmake_binary="$(find_cmake_version 3 25 0 2>/dev/null)"
+#
+# Return 0 (true) if a CMake binary matching the requested version is available.
+# Return a non-zero (false) value otherwise.
+#
+# If successful, print the name of the CMake binary to stdout (pipe 1).
+# Otherwise, no output is printed to stdout (pipe 1).
+#
+# Diagnostic messages may be printed to stderr (pipe 2). Redirect to /dev/null
+# with `2>/dev/null` to silence these messages.
+#
+# Example:
+#   cmake_binary="$(find_cmake_version 3 25 0)" || exit
+#   "${cmake_binary:?}" -S path-to-source -B path-to-build
+find_cmake_version() {
+  declare -r major="${1:?"missing CMake major version"}"
+  declare -r minor="${2:?"missing CMake minor version"}"
+  declare -r patch="${3:?"missing CMake patch version"}"
+  declare -r version="${major}.${minor}.${patch}"
+
+  declare -r cache_dir="$HOME/.cache/mongo-c-driver"
```

Comment:
Recommend using a more thorough cache path definition, especially one to share across scripts and use cases: e.g. https://github.com/mongodb/libmongocrypt/blob/1d89e986fe682af4b3b8244014f2a6b62dc0b7b0/.evergreen/init.sh#L235-L256

This allows rerouting caches (e.g. into a container bind mount or a `/tmp` dir) and busting the cache more reliably should things go awry.

Reply:
`XDG_CACHE_HOME` is a control point for users to redirect application caches, so it is rarely set by a distribution. Similar story for redirecting the cache in general: It's not for EVG specifically, but for downstream developers or wrapping scripts that use these scripts within their own system and want to control the caching behavior.

The purpose for keying the cache (such as with a number) is to permit multiple versions to simultaneously exist on a system. e.g. Build $A$ uses key `a`, but $B$ uses key `b`. $B$ and $A$ will separately have idempotent caching without wiping each other's out. This is useful when performing changes that could make the cache contents incompatible, but builds between them might interleave on a single host during the transition.

Reply:
Although I understand the use cases described, I remain unconvinced that this is not deep in [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) territory. Supporting "downstream developers" (who?) or "wrapping scripts that use these scripts within their own system and want to control the caching behavior" (again, who?) seems far beyond the scope of these scripts' intended purpose, which is to ensure a CMake binary fulfilling the requested version is present. The use of a cache is just to avoid extra work if able.

> idempotent caching without wiping each other's out

The design and definition of `cmake_replace_version`, specifically the use of `cmake-${version}` as a _symlink_ to the actual CMake root directory (which is also present in the cache), is supposed to address this concern. Given the requirement being fulfilled by this script is just to provide a CMake binary with the requested version, it does not matter which CMake root directory a given call to CMake uses so long as the CMake binary is executable and behaves correctly.

Multiple CMake root directories may be installed in parallel within the mongo-c-driver cache (a bit of unnecessary work), but this process does not interfere with each other due to the use of `make_tmpdir_in` (each uses its own CMake root directory within the cache). There is only a single point of potential contention, is the atomic update of the symlink through which the CMake binary is executed. Old CMake root directories within the cache remain completely untouched. If the symlink exists, by construction (as defined by the `find_cmake_version` function), it should always point to a valid CMake binary within a valid CMake root directory.